### ISSUE TITLE
LLM Feature Request: logprobs

### ISSUE THREAD

**Issue by bgoelTT** [tenstorrent/tt-metal#32210]
The `logprobs` parameter in the OpenAI chat completions API https://platform.openai.com/docs/api-reference/chat/create#chat_create-logprobs signals the model to return log probabilities for each output token. We need to enable the configuration of the logprobs parameter in our on-device sampling implementation, for each request in a batch. Note that the models should optionally return logprobs - with the caller making use of them if specified.

This issue serves as the parent issue for all tasks which are needed to expose this parameter.

| Model           | Log-Probs Enabled | Avg TSU | Time  |
|-----------------|-----------------|---------|-------|
| Qwen3 T3K       | No              |     22.14 t/s/u    |   45.16ms    |
| Qwen3 T3K       | Yes             |     16.29 t/s/u    |   61.40ms    |
| Llama3.1-8B     | No              |     48.83 t/s/u    |   20.48ms    |
| Llama3.1-8B     | Yes             |   29.75 t/s/u  |   33.61ms   |

**Comment by sraizada-tt** [comment]
@tt-ai can you make sub-issues for this issue with:
1. add optional returning of log-softmaxed logins from model 
2. pass get_log-probs from vllm to model forward functions

**Comment by djordje-tt** [comment]
We are unfortunately unable to calculate precise logprobs with current sampling op since we are taking only top_k=32 per device and all further calculations base only on 32*8=256 values which dismisses all other values (vocab_size - 256) and softmax does not provide completely correct values. 

By looking in industry standards I’ve seen that it’s common practice to do top-k, gather, and then softmax and further processing for both log-probs and calculating next token so I think it’s fine that we also do the same thing. 

Another option is to add additional step that takes all vocab_size logits, executes softmax and log and returns correct log-prob. This would limit couple of things: 
- perf would be decreased (don't have estimation now) but if it's debug-only feature I believe it should be fine.
- vLLM requests could only be with or without log-probs flag otherwise we would need multiple traces where one contains additional step and another does not.

@bgoelTT what are your thoughts on this? 


In the meantime, I will start implementing option with additional step where we calculate precise log-prob.

**Comment by ekuslysTT** [comment]
Notes
- mimic vllm

**Comment by ekuslysTT** [comment]
Notes
- integrate with TT transformers. Potentially creating to support subcore grid so we can run on llama 70. --> check if works fine on single core, if not open issue and assign to @cmaryanTT  team.

**Comment by ekuslysTT** [comment]
Notes:
1. accuracy issue: after accuracy improvement PR was merged, log probs were zero.  Distribution such that 1 token is highly probably all others are insignificant. Output should be very small due to using bfloat16 but gets truncated to zero. No improvement with float16. Does not affect most probably token --> not blocking integration work
2. llama3.1-8B: logprobs were off, between -20;-30 rather than 0;1. fixed --> waiting for vllm nightly to pass.

vllm: nightly tests failed, debugging WIP @rdraskicTT

**Comment by ekuslysTT** [comment]
Notes
- working for TTT
- blocked by PR for vllm
- on tt-metal side working for both llama and qwen
- - waiting with merge until it's passing in order not to affect other pipelines

Action Items:
- once both PRs are merged @djordje-tt to share PRs with @bgoelTT

**Comment by ekuslysTT** [comment]
@djordje-tt create new subissue

**Comment by ekuslysTT** [comment]
@djordje-tt assess work by tomorrow 12/10/25, pull in more people if needed.

**Comment by djordje-tt** [comment]
Found workaround for ongoing bugs and running Models CI to check if it's passing with logprobs. Once it's passing we can close it for TT-Transformers.

**Comment by ekuslysTT** [comment]
Notes
- model Ci tests are passing except the one for the penalties that was already failing earlier. https://github.com/tenstorrent/tt-metal/issues/34090
- waiting for nightly CI run

**Comment by ekuslysTT** [comment]
Sub-issues https://github.com/tenstorrent/tt-metal/issues/34077 and https://github.com/tenstorrent/tt-metal/issues/34080 are P1 for Q1'26

**Comment by ekuslysTT** [comment]
Blocked by logprobs enablement on Galaxy-optimized models https://github.com/tenstorrent/tt-metal/issues/34161

**Comment by ekuslysTT** [comment]
Closing after 12/31 release

**Comment by ekuslysTT** [comment]
@djordje-tt and @bgoelTT this is marked as a open P0. Are the remaining sub-issues still release-critical?
